### DISCUSSION TITLE
deepMap can alter structure?

### DISCUSSION THREAD

**Discussion by masciugo** [YuriGor/deepdash#97] [Q&A]
I used deep map to integrate some object node with extra properties and deepMap works perfectly returning new value where I want. The structure of the mapped object remains the same. 

Let's suppose I need to transform array of strings node in a leaf with those string concatenated altering the structure:

{ arr: ['a', 'b', { c: ['y', 'w', 'z'] }, 'd', 'e'] }

{ arr: ['a', 'b', { c: "ywz" }, 'd', 'e'] }

I was no able to do that with deepMap. Is it possible?

**Comment by YuriGor** [comment]
Probably i will need to enhance mapDeep, right now it's possible to do with eachDeep, but i am out of my laptop, so i will be able to create an example only in some hours.

**Comment by YuriGor** [comment]
I hope I understood your criteria right:
you want to replace any flat array by CSV string?
here is an example:
```js
var data = { arr: ['a', 'b', { c: ['y', 'w', 'z'] }, 'd', 'e',[1,2,3],{more:{nested:{arrays:[[1,2,3],[4,5,6],[7,8,9]]}}}]};
_.eachDeep(data,(val, key, parent, context)=>{
  if(_.isArray(val) && !_.some(val, _.isObject)){
    parent[key] = val.join();
    return false;
  }
});
```
data:
```json
{
  "arr": [
    "a",
    "b",
    {
      "c": "y,w,z"
    },
    "d",
    "e",
    "1,2,3",
    {
      "more": {
        "nested": {
          "arrays": [
            "1,2,3",
            "4,5,6",
            "7,8,9"
          ]
        }
      }
    }
  ]
}
```
[play with demo here](https://codepen.io/yurigor/pen/zYYMRbq?editors=0010)

**Comment by YuriGor** [comment]
And about topic question: "can mapDeep alter structure?"
As said [in docs](https://deepdash.io/#mapdeep)
"mapDeep returns an object with the **same structure** with values trasformed thru iteratee."
So, no, it cannot. 
And I will keep it as is because it's made intentionally, to mimic the behavior of native `Array.map` method (And it's not cargo-cult - it really makes sense).

Let me know please, if my suggestion works for you.
